services/src/main/java/com/chatbot/services/ChatServiceControllers/HangoutsController.java

```diff
@@ -0,0 +1,179 @@
+import java.util.List;
+import java.util.Map;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.api.services.chat.v1.model.ActionResponse;
+import com.google.api.services.chat.v1.model.Card;
+import com.google.api.services.chat.v1.model.Message;
+import com.google.api.services.chat.v1.model.Section;
+import com.google.api.services.chat.v1.model.TextParagraph;
+import com.google.api.services.chat.v1.model.WidgetMarkup;
+
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.chatbot.services.ChatServiceConstants;
+import com.chatbot.services.IDMapping;
+import com.chatbot.services.AsyncServices.HangoutsAsyncService;
+import com.chatbot.services.protobuf.ChatServiceRequestOuterClass.ChatServiceRequest;
+import com.chatbot.services.protobuf.ChatServiceRequestOuterClass.ChatServiceRequest.ChatClient;
+import com.chatbot.services.protobuf.ChatServiceRequestOuterClass.ChatServiceRequest.MimeType;
+import com.chatbot.services.protobuf.ChatServiceRequestOuterClass.ChatServiceRequest.RequestType;
+
+@RestController
+public class HangoutsController extends ChatServiceController {
+
+  private HangoutsAsyncService asyncService;
+  private HangoutsAuth hangoutsAuth;
+  
+  private static final String REMOVED_FROM_SPACE_EVENT = "REMOVED_FROM_SPACE";
+  private static final String ADDED_TO_SPACE_EVENT = "ADDED_TO_SPACE";
+  private static final String MESSAGE_EVENT = "MESSAGE";
+  private static final String CARD_CLICKED_EVENT = "CARD_CLICKED";
+
+  // Receive messages from hangouts and dispatch response calls
+
+  public HangoutsController(HangoutsAsyncService asyncServiceToSet,
+      HangoutsAuth hangoutsAuthToSet) {
+    asyncService = asyncServiceToSet;
+    hangoutsAuth = hangoutsAuthToSet;
+  }
+
+  @Override
+  @PostMapping("/")
+  Message onRequest(@RequestHeader final Map<String, String> headers,
+      @RequestBody final JsonNode event) throws Exception {
+    if (headers.get("user-agent").equals(ChatServiceConstants.HANGOUTS_USER_AGENT)) {
+      hangoutsAuth.verifyRequest(headers);
+      asyncService.chatServiceRequestHandler(buildChatServiceRequestFromHTTPRequest(event));
+    } else {
+      throw new IllegalArgumentException("Unknown user agent");
+    }
+    final Message reply = new Message();
+    if (!event.at("/type").asText().equals(CARD_CLICKED_EVENT)) {
+      // an empty message acts as an acknowledgement for a text message
+      reply.setText("");
+    } else {
+      // replace the button with a text paragraph to render it unclickable
+      reply.setActionResponse((new ActionResponse()).set("type", "UPDATE_MESSAGE"));
+      final List<WidgetMarkup> widgets = new ArrayList<>();
+      final TextParagraph widget =
+          new TextParagraph().setText(event.at("/action/parameters/0/value").asText());
```

Comment:
for all static const values, define it at class level

Reply:
updated

Reply:
I don't see any update for this

Reply:
The variables were only used in one place so instead of making them class variables, moved generation of textparagraph card to a function.